### ISSUE TITLE
Display the number of running docker (other?) containers

### ISSUE THREAD

**Issue by milouse** [nojhan/liquidprompt#781]
Allow the user to remember about running containers locally, or give a quick hint about possible issue (if none is shown and you know some should be there…)

### Description

Add a new item in the LP_JOBS category showing the current number of running docker containers.

### How will this be useful?

As a developer, docker may now be part of the day to day job. Very often we start dev containers to test things independently from our current environment. That’s quick and efficient.

However docker itself may be a bit resource consuming, or we might forgot stuff (like when you start detached containers and forget about that database).

As a little system admin (home hosting), some product are often simpler to run as docker containers than trying to install them by hand (mastodon, mobilizon…). It might be interesting to have a quick hint to ensure every thing looks fine if you already know that 3 containers should be running).

### Example prompt

[milouse: ~ ] $ sudo docker start mariadb
[milouse: ~ ] 1🐳 $ nano test.txt

Utilisez « fg » pour revenir à nano.
zsh: suspended (signal)  nano test.txt
[milouse: ~ ] 1z 1🐳 147 $ uname -mn
lupin x86_64
[milouse: ~ ]
1z 1🐳 $

### Implementation details

The problem with docker toolchain is that it very often require elevated permissions. Following the system, one may have to use sudo, or be root. Thus directly calling `docker ps` does not look to be a good idea. It was my first approach still, by ensuring that sudo docker is allowed without password prompt for my current user.

```bash
local containers="$(sudo docker ps -q); printf x"
local -i count
__lp_count_line "${containers%x}"
lp_running_docker_containers="$count"
```

Thus I was looking for another, sudo-less approach, and may have found a low-level/read only solution for latest docker version:

```bash
# Return the count of detached running docker containers
_lp_docker_container_count() {
    (( LP_ENABLE_DOCKER_CONTAINERS )) || return 2
    # I’ve trouble to deal with glob name, zsh, and in the end I need to
    # count (grep -c, not just grep), thus it is not that bad to keep the ls | grep
    # shellcheck disable=SC2010
    lp_running_docker_containers="$(\ls -1 /sys/fs/cgroup/system.slice | GREP_OPTIONS='' \grep -c "docker-.*\.scope")"
}

_lp_jobcount_color () {
    (( LP_ENABLE_JOBS || LP_ENABLE_DETACHED_SESSIONS || LP_ENABLE_DOCKER_CONTAINERS )) || return 2

    # …

    if _lp_docker_container_count; then
        if (( lp_running_docker_containers > 0 )); then
            [[ -n "$lp_jobcount_color" ]] && lp_jobcount_color+="$LP_MARK_JOBS_SEPARATOR"
            # ⁙❒⚙🐳
            lp_jobcount_color+="${LP_COLOR_JOB_R}${lp_running_docker_containers}🐳${NO_COL}"
        fi
    fi
    # …
}
```

I already integrated it in my own them and it’s working pretty well.

What do you think of it?

If it looks interesting, it will need some time to [support all docker version](https://docs.docker.com/config/containers/runmetrics/#find-the-cgroup-for-a-given-container)

**Comment by Rycieos** [comment]
This is 100% a good idea. If a working PR lands in my inbox, I'll merge it no questions (provided it has docs and such of course).

Your implementation is a decent start (especially your first approach, which is exactly what I was thinking), but requires some more thought. Here are my questions:
1. `/sys/fs/cgroup/system.slice` must be a new kernel feature. My 4.18 kernel does not have that file, while my 5.14 kernel does. We should support more than just the latest kernels.
2. `docker` can be run as non-root pretty easily by using the built-in `docker` group feature. Just add your user to the `docker` group, and you can run any docker command. I would rather recommend that method than try to use `sudo`.
3. Liquid Prompt now supports detecting if its shell is running _inside_ a container. We detect the types Singularity, Toolbox, Podman, Docker, LXC, and systemd. Any suggested "jobs" extension should attempt to support as many of these types as possible.
4. Do we want to lump all types of containers together? It is possible to run Docker and systemd containers on the same system, and probably possible to do the same with others. If we can find an interface directly on the kernel to get containers, it would simplify the approach and have the benefit of supporting any new container engines as well.

**Comment by milouse** [comment]
Hi,

Yes it was more or less a proof of concept. But I’m glad you like it. I’ll try to improve it and once I feel it’s ready I’ll open a PR.

Some answers to your question (but discussion is still open I think, I’m absolutely not an expert in this domain, only a user):

1. Yep, I confirm everything depends upon the kernel version and also the usage of systemd or not (and which version). On the docker documentation itself they already speaks about 4 different paths, but there are probably more.
2. The documenation of my distribution discourages the use of the `docker` group by pointing to [this upstream issue](https://github.com/moby/moby/issues/9976) (more a discussion than an issue). The point is that the docker daemon need root priviledges and is able to run highly priviledged code and thus, allowing any user to blindly run docker command is not a good idea. Yes exactly like what I exposed with my own sudo config… So it means it is up to anyone to do whatever they want on their machine, but from a security point of view those path should not be blindly advised. This is at least how I understand it. For the rest I continue in point 4.
3. I agree… but as I do not use any of them, in any case if I open a PR it will not support any of them. But I can make it the most generic possible to allow gradual improvement.
4. This is the second reason (continuing from the point 2.) why I finally switched to a solution inspecting the `/sys` filesystem. Not only it removes the security issue, but it also sounds more generic/extendable for other type of cgroup-related containers.

Mash-up together, it makes me think about some kind of possible fallback mechanism: 1. try to support kernel interface to avoid security issue and improve performance (I think counting file will always be quicker than calling docker or anything else bin) 2. if this is not possible (old kernel, specific configuration…) fallback to docker binary or else.

**Comment by Rycieos** [comment]
The fallback mechanism is what we now do for things like temperature that can be gathered with the `/sys` interface, since it is much faster. I agree it is the right approach.

**Comment by nojhan** [comment]
Wouldn't this be useful with a more general approach? Like answering a question such as "is this currently running on my machine?". 

Like checking if any service is up ("is my database running?"), not just containerized ones?
I guess that if one parses `/sys/`, it should be feasible to grep any process.

If containers need some specific parsing config, at least having a common display segment would be nice.

**Comment by Rycieos** [comment]
> Wouldn't this be useful with a more general approach? Like answering a question such as "is this currently running on my machine?".

Yes, probably. The problem is that Liquid Prompt just isn't built to handle general cases. We keep adding features that push it toward that ideal, and my work with the v2 themes has made it more like a "prompt theme engine", but it has a long way to go; the user interface for configuring such features is clunky. And probably Shell is not the right language to do that in.

I have rising and falling interest in building such a "prompt theme engine", but it would take a lot of work.
